examples/step-74/step-laplace-FEIV.cc

```diff
@@ -0,0 +1,1159 @@
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the deal.II distribution.
+ *
+ * ---------------------------------------------------------------------
+
+ *
+ * Author: Timo Heister and Jiaqi Zhang, Clemson University, 2020
+ */
+
+
+// The first few files have already been covered in previous examples and will
+// thus not be further commented on:
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/function.h>
+#include <deal.II/base/function_lib.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/sparse_direct.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/fe/mapping_q1.h>
+// Here the discontinuous finite elements are defined. They are used in the
+// same way as all other finite elements, though -- as you have seen in
+// previous tutorial programs -- there isn't much user interaction with finite
+// element classes at all: they are passed to <code>DoFHandler</code> and
+// <code>FEValues</code> objects, and that is about it.
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_interface_values.h>
+
+#include <deal.II/lac/solver_gmres.h>
+#include <deal.II/lac/precondition_block.h>
+#include <deal.II/lac/precondition.h>
+
+#include <deal.II/numerics/derivative_approximation.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/base/convergence_table.h>
+
+#include <deal.II/meshworker/copy_data.h>
+#include <deal.II/meshworker/mesh_loop.h>
+#include <deal.II/meshworker/scratch_data.h>
+
+
+
+namespace Step74
+{
+  using namespace dealii;
+
+  enum Test_Case
+  {
+    convergence_rate,
+    l_singularity
+  };
+
+
+  template <int dim>
+  class Solution : public Function<dim>
+  {
+  public:
+    Solution(Test_Case test_case)
+      : Function<dim>()
+      , test_case(test_case)
+    {}
+    virtual void           value_list(const std::vector<Point<dim>> &points,
+                                      std::vector<double> &          values,
+                                      const unsigned int             component = 0) const;
+    virtual Tensor<1, dim> gradient(const Point<dim> & point,
+                                    const unsigned int component = 0) const;
+
+  private:
+    Functions::LSingularityFunction ref;
+    Test_Case                       test_case;
+  };
```

Comment:
Right now the logic is to have a single analytical solution (and RHS) for all the cases. The difference between the cases is set by the Test_Case enum. This is works well.
I think it would be cleaner to have one class for each problem and to store a pointer to the solution in the solver class. Then depending on the test case, you choose which solution to instantiate. This prevents the test_case for percolating everywhere. This is only an opinion, feel free to do as you will :)

Reply:
That's a good idea, but I am not sure how to implement it. Could you provide an simple example, or which deal.ii tutorial has implemented something like this?